### PR TITLE
Add support for CWT Claims & Type in Protected Headers

### LINKED ISSUE #173: Support for CWT Claims in Headers
### What is the areas you would like to add the new feature to?

Go-COSE Library

### Is your feature request related to a problem?

See https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/

### What solution do you propose?

See https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/

### What alternatives have you considered?

None.

### Any additional context?

Once this document becomes an RFC it is advisable to add support for it to protected headers.

### LINKED ISSUE #184: Support for typ in protected header
### What is the areas you would like to add the new feature to?

Go-COSE Library

### Is your feature request related to a problem?

I want to be able to use the `typ` header, as described in https://datatracker.ietf.org/doc/draft-ietf-cose-typ-header-parameter/

### What solution do you propose?

Add the code point for the `typ` parameter to the header parameters exposed by this library.

### What alternatives have you considered?

setting the parameter directly, without using the header parameters exposed by this library.

### Any additional context?

This should be done after https://datatracker.ietf.org/doc/draft-ietf-cose-typ-header-parameter/ has been assigned an RFC number.

### PR DESCRIPTION
**Pull request by OR13**
closes #173

closes #184 

This PR should remain open until numbers are assigned to:

- https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/
- https://datatracker.ietf.org/doc/draft-ietf-cose-typ-header-parameter/

### CODE REVIEW

File: cwt_test.go
Comment (OR13) [inline cwt_test.go:52]:
```suggestion
```

Comment (OR13) [reply cwt_test.go:52]: I committed this, so the reviewers can inspect the diagnostic, but it should be removed before merging.

File: headers.go
Comment (OR13) [inline headers.go:108]: It could be argued this is essential to do, before merging this.

File: cwt.go
Comment (OR13) [inline cwt.go:15]: I don't think it is smart to add the whole registry in the first PR.

File: cwt_test.go
Comment (OR13) [inline cwt_test.go:9]:
```suggestion
```

File: cwt_test.go
Comment (OR13) [inline cwt_test.go:74]: for this test, it might be better to tamper with the protected header.

**Comment by SteveLasker** [comment]
@qmuntal, @shizhMSFT, @thomas-fossati, @henkbirkholz can you please take a look, and provide some feedback to @OR13

**Review by shizhMSFT — COMMENTED** [review]
Does `CWT` deserve its own package `package cwt` under `package cose`?

**Comment by SteveLasker** [comment]
**go-cose call feedback**: concesus was to do it in this package.

File: cwt.go
Comment (SteveLasker) [reply cwt.go:15]: Based on the above, it looks like we have the minimal code needed to proceed.

Comment (yogeshbdeshpande) [reply cwt.go:15]: We should add what is bare minimum required, and then extend it when a new point needs to be registered!

File: headers.go
Comment (SteveLasker) [reply headers.go:108]: Proposing to track as a separate issue, rather than block this PR. See #186

**Comment by SteveLasker** [comment]
blocked, awaiting: https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/

**Comment by SteveLasker** [comment]
https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/ is [progressing](https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/history/) is progressing. Expectation a few more weeks.

**Review by henkbirkholz — APPROVED** [review]
correctly reflects:
https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/10/
which will move with:
https://datatracker.ietf.org/doc/draft-ietf-cose-typ-header-parameter/05/

**Review by SteveLasker — CHANGES_REQUESTED** [review]
Approve of the changes, however we decided to wait for https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/ to be completed, with a comment added referencing the RFC #.

**Comment by OR13** [comment]
I added support for the `typ` header parameter:

- https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/
- https://datatracker.ietf.org/doc/draft-ietf-cose-typ-header-parameter/

```
18([
h'a30138230fa2016e6973737565722e6578616d706c65026f7375626a6563742e6578616d706c65106f6170706c69636174696f6e2f637774', 
{4: h'31'}, 
h'68656c6c6f20776f726c64', h'00b8f0fbee7d5ca003678a173a1f1cb5f6233e85f2dd421d4d56d13541a67e01fd0c7addabc2c3b07d5b08a027382629aac41dd3f62f69d6c9209e2ca99255bda78600f3ca500d482c9b3a220dfd14395c42d02e0fa423e9192deb83c27b41257bcb3e9162db9e3de0895a81fb51d2ae41e9f07d91146832cbe91fa85b48456aef8ebc82'
])
```

**Comment by OR13** [comment]
I suppose there is an API consideration here, perhaps the validation checks should only happen when serialization occurs, instead of causing an error to be thrown when for example invalid claims or invalid cose type is set.

**Comment by SteveLasker** [comment]
Sooooooo, close.

File: cwt.go
Comment (thomas-fossati) [inline cwt.go:3]:
```suggestion
// https://www.iana.org/assignments/cwt/cwt.xhtml#claims-registry
```

File: cwt.go
Comment (thomas-fossati) [reply cwt.go:15]: > [...] and then extend it when a new point needs to be registered!

Sorry, I have probably missed the right conversations, but I am not sure I get what the strategy is.  There are 10's of claims already registered.  What is the criterion for promoting them from the [CWT Claims](https://www.iana.org/assignments/cwt/cwt.xhtml#claims-registry) registry to `const`s in the `cose` package?

File: cwt_test.go
Comment (thomas-fossati) [inline cwt_test.go:14]: spurious NL

```suggestion
```

File: cwt_test.go
Comment (thomas-fossati) [inline cwt_test.go:46]: awesome :)

File: headers.go
Comment (thomas-fossati) [inline headers.go:517]: this looks worth refactoring (see also L524-537).

**Review by thomas-fossati — COMMENTED** [review]
marvellous, thanks!

I have left a few (mostly silly) comments for your consideration.

File: cwt_test.go
Comment (thomas-fossati) [inline cwt_test.go:23]: more idiomatic:

```suggestion
	claims := cose.CWTClaims{
		cose.CWTClaimIssuer:  "issuer.example",
		cose.CWTClaimSubject: "subject.example",
	}
```

**Comment by OR13** [comment]
@thomas-fossati regarding https://github.com/veraison/go-cose/pull/183#discussion_r1649136571

Its a fair point. 

My lazy answer is, people can add pull requests for the IANA registry entries they wish the library supported, and they can use the registries without adding "developer experience for them", even if we don't support them here.

We could decide to not add any registry entries from the CWT registry, and only support the registered headers.

File: headers.go
Comment (OR13) [reply headers.go:517]: I agree, but I am not really a go dev... probably better for someone with better intuition to attempt it

**Comment by OR13** [comment]
The combination of a cross fork PR, and the DCO rebase + github suggestion merges... has made this more trouble than its worth.

@thomas-fossati If you think we should drop the CWT side of this, I will probably just do a fresh PR that only defines the header parameters, and provide some examples that show how to use them, without defining any CWT registry specific stuff.

**Review by SteveLasker — APPROVED** [review]

File: headers.go
Comment (thomas-fossati) [reply headers.go:517]: No worries, we can refactor it later.

**Comment by thomas-fossati** [comment]
> The combination of a cross fork PR, and the DCO rebase + github suggestion merges... has made this more trouble than its worth.
> 
> @thomas-fossati If you think we should drop the CWT side of this, I will probably just do a fresh PR that only defines the header parameters, and provide some examples that show how to use them, without defining any CWT registry specific stuff.

I believe this PR is good and worth adding into the mainline as-is.

Let's raise a couple of tracking issues (one for the registration, one for the potential refactoring) and let's get it merged.

Thanks again for the great contribution.

**Review by thomas-fossati — APPROVED** [review]
🚢 it

**Comment by SteveLasker** [comment]
@OR13, did you want to re-submit this as a new/clean PR, or can we merge as-is?

With #187 and #188 queued up, we're ready for a new release.

**Comment by SteveLasker** [comment]
Per discussion with @OR13, it was easier to close and redo this PR than fix the DCO issues.  
Closing as this with replaced PR #189

**Pull request closed without merging**
